File: data_structures/arrays_and_strings/str_compressor.rb
def str_compressor(str)
  chars = str.split('')
  compressed_str = ''
  count = 1
  chars.each_with_index do |el, idx|
    if el == chars[idx + 1]
      count += 1
    else
      compressed_str << el + count.to_s
      count = 1
    end
  end
  compressed_str.length < str.length ? compressed_str : str
end

# Time complexity: O(n)
# Space complexity: O(n)

File: data_structures/arrays_and_strings/str_compressor.js
function strCompressor(str) {
  let chars = str.split("");
  let compressedStr = "";
  let count = 1;
  for (let i = 0; i < chars.length - 1; i++) {
    if (chars[i] !== chars[i + 1]) {
      compressedStr += chars[i] + count;
      count = 1;
    } else if (i === chars.length - 2) {
      count += 1;
      compressedStr += chars[i] + count;
    } else {
      count += 1;
    }
  }
  return compressedStr.length < str.length ? compressedStr : str;
}

// Not the most elegant solution for dealing with that last letter, may revisit later